### PR TITLE
FIX: tickspacing for subfigures

### LINKED ISSUE #20765: savefig re-scales xticks and labels of some (but not all) subplots
The x-ticks (and respective labels) for some of my subplot axes are re-scaled in the output png files.

image

While I didn't find this exact problem posted elsewhere, perhaps it's related to:
https://github.com/matplotlib/matplotlib/issues/12564/

I say that b/c, the above image was my standard export at 300 dpi while I get my expected result at 120 dpi (below):

image

Changing the figsize changes the expression of the problem, but doesn't make it go away (e.g. 300 dpi at 14 x 20):

image

I also consulted :
[(https://github.com/matplotlib/matplotlib/issues/6800)]
[(https://github.com/matplotlib/matplotlib/issues/6638)]

but they are different expressions (of possibly a similar underlying problem) than mine.

I find it particularly odd that the minor x-ticks seem to stay put while the major x-ticks (and respective labels) shift.

Minimally reproducible code (with min/maxima of x-values in my data series plotted as points):
```
fig = plt.figure(constrained_layout=True, figsize=(10, 14))

# create top/bottom subfigs
(subfig_t, subfig_b) = fig.subfigures(2, 1, hspace=0.01, height_ratios=[3, 7])

# put ax0 in top subfig
ax0 = subfig_t.subplots()

# create left/right subfigs nested in bottom subfig
(subfig_bl, subfig_br) = subfig_b.subfigures(1, 2, wspace=0.01, width_ratios=[7, 2])

# put ax1-ax3 in gridspec of bottom-left subfig
gs = subfig_bl.add_gridspec(nrows=1, ncols=14)

ax1 = subfig_bl.add_subplot(gs[0, :1])
ax1.scatter(x=[-56.46881504821776, 24.179891162109396], y=[1500, 3600])


ax2 = subfig_bl.add_subplot(gs[0, 1:3], sharey=ax1)
ax2.scatter(x=[-126.5357270050049, 94.68456736755368], y=[1500, 3600])

ax3 = subfig_bl.add_subplot(gs[0, 3:14], sharey=ax1)
ax3.scatter(x=[-638.323228561402, 878.5568315887451], y=[1500, 3600])

ax2.set_ylabel('')
ax2.set_yticklabels([])
ax3.set_ylabel('')
ax3.set_yticklabels([])

b1_xlab = subfig_bl.supxlabel("MyLabel", x=0.5, size=12, fontweight='bold')
b1_xlab._autopos = True

# put ax4 in bottom-right subfig
ax4 = subfig_br.subplots()
ax4.scatter(x=[-36.00010414731702, 46.8026446716616], y=[1500, 3600])

subfig_br.supxlabel(' My Other Label', size=12, fontweight='bold')

dpi_lst = [120, 300]

for d in dpi_lst:
    out_plot = out_path + "MyFile_{}.png".format(d)
    plt.savefig(out_plot, dpi=d, transparent=True)    
```

Any suggestions for a work-around?

Operating system
Windows 10 Pro

Matplotlib Version
3.4.2

Python version
3.9.5

Jupyter version
3.0.16

Installation
conda

Conda channel
conda-forge

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #20765

tick spacing was incorrect for subfigure.  The issue was pretty subtle, but the dpi transform was not correct in `x/yaxis.get_tick_spacing` when there were subfigures.

The new test fails on master with, and you can see that 300 dpi has different list of ticks than 120 dpi

```
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0
E       
E       (shapes (5,), (7,) mismatch)
E        x: array([-200., -100.,    0.,  100.,  200.])
E        y: array([-150., -100.,  -50.,    0.,   50.,  100.,  150.])

lib/matplotlib/tests/test_figure.py:1058: AssertionError
```

The test is a bit convoluted, but that is because it is tricky to get the ticks to fall right on the boundary where the difference can be seen.

### CODE REVIEW

**Comment by jklymak** [comment]
Again, prob not urgent for 3.4.3...

**Comment by anntzer** [comment]
Should we check for all instances of `.figure.dpi` in the codebase and test whether they need a similar fix?

**Comment by jklymak** [comment]
Lets move to draft and I'll do a census as @anntzer suggests.  

As a note, I notice that the new test also makes the markersize change with dpi, which I don't think happens with normal figures.

**Comment by jklymak** [comment]
A quick scan doesn't show any other cases.  Note `figure.dpi` appears a lot, but this was an axes-coords->pixels->points conversion.  Most dpi conversions are pixels->points or pixels->inches, which don't care about the axes co-ordinates.

File: lib/matplotlib/tests/test_figure.py [hunk 1]
```diff
@@ -1031,6 +1031,34 @@ def test_subfigure_spanning():
     np.testing.assert_allclose(sub_figs[2].bbox.max, [w, h / 3])
 
 
+@mpl.style.context('mpl20')
+def test_subfigure_ticks():
+    # This tests a tick-spacing error that only seems applicable
+    # when the subfigures are saved to file.  Its very hard to replicate
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:1037]:
```suggestion
    # when the subfigures are saved to file.  It's very hard to replicate.
```

**Review by QuLogic — APPROVED** [review]

**Comment by jklymak** [comment]
See #20777  for another case.

**Comment by jklymak** [comment]
This now includes a setter/getter property for `SubFigure.dpi` the redirects to the parent.  Other properties on the SubFigure that are inherited from the parent are already pointers....

File: lib/matplotlib/tests/test_figure.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/tests/test_figure.py:1037]: This appears to have undone.

Comment (jklymak) [reply lib/matplotlib/tests/test_figure.py]: Should be fixed now...

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -2001,6 +2000,14 @@ def __init__(self, parent, subplotspec, *,
         if parent._layoutgrid is not None:
             self.init_layoutgrid()
 
+    @property
+    def dpi(self):
+        return self._parent.dpi
+
+    @dpi.setter
```
Comment (tacaswell) [inline lib/matplotlib/figure.py:2007]: not sure we want the setter here?  The getter makes sense as we may need this in the transforms, but do we want to be able to change the dpi of the whole figure from a sub figure?

On the other hard it seems realtively harmless and makes sure that subfigures feel as like figures as they can.

**Review by tacaswell — APPROVED** [review]

File: lib/matplotlib/figure.py (hunk 2 above)
Comment (QuLogic) [reply lib/matplotlib/figure.py:2007]: I actually tried it, and it breaks some tests without the setter. I did not investigate any further though, so I don't know if that's a real problem or not.

Comment (jklymak) [reply lib/matplotlib/figure.py:2007]: Happy to mess around with that.  I'd say dpi is one of our more fragile concepts.

**Pull request merged**